src/curve/constant-product/ConstantProductPair.sol

```diff
@@ -248,72 +167,110 @@ contract ConstantProductPair is IConstantProductPair, UniswapV2ERC20, AssetManag
-            balance1 = _totalToken1();
+        uint256 amountIn;
+        address tokenOut;
+
+        // exact in
+        if (inOrOut) {
+            // swap token0 exact in for token1 variable out
+            if (amount > 0) {
+                tokenOut = token1;
+                amountIn = _totalToken0() - _reserve0;
+                require(amountIn == uint256(amount), "CP: AMOUNT_MISMATCH");
+                amountOut = _getAmountOut(amountIn, _reserve0, _reserve1, swapFee);
+            }
+            // swap token1 exact in for token0 variable out
+            else {
+                tokenOut = token0;
+                amountIn = _totalToken1() - _reserve1;
+                require(amountIn == uint256(-amount), "CP: AMOUNT_MISMATCH");
+                amountOut = _getAmountOut(amountIn, _reserve1, _reserve0, swapFee);
+            }
         }
-        uint amount0In = balance0 > _reserve0 - amount0Out ? balance0 - (_reserve0 - amount0Out) : 0;
-        uint amount1In = balance1 > _reserve1 - amount1Out ? balance1 - (_reserve1 - amount1Out) : 0;
-        require(amount0In > 0 || amount1In > 0, "CP: INSUFFICIENT_INPUT_AMOUNT");
-        { // scope for reserve{0,1}Adjusted, avoids stack too deep errors
-            uint balance0Adjusted = (balance0 * 10000) - (amount0In * swapFee);
-            uint balance1Adjusted = (balance1 * 10000) - (amount1In * swapFee);
-            require(balance0Adjusted * balance1Adjusted >= uint(_reserve0) * _reserve1 * (10000**2), "CP: K");
+        // exact out
+        else {
+            uint256 actualAmountIn;
+            // swap token1 variable in for token0 exact out
+            if (amount > 0) {
+                tokenOut = token1;
+                amountOut = uint256(amount);
+                amountIn = _getAmountIn(amountOut, _reserve1, _reserve0, swapFee);
+                actualAmountIn = _totalToken1() - _reserve1;
+            }
+            // swap token0 variable in for token1 exact out
+            else {
+                tokenOut = token0;
+                amountOut = uint256(-amount);
+                amountIn = _getAmountIn(amountOut, _reserve0, _reserve1, swapFee);
+                actualAmountIn = _totalToken0() - _reserve0;
+            }
+            require(amountIn <= actualAmountIn, "CP: INSUFFICIENT_AMOUNT_IN");
+            if (amountIn < actualAmountIn) {
+                // refund the user if the actualAmountIn is too much
+                _safeTransfer(tokenOut == token0 ? token1 : token0, to, actualAmountIn - amountIn);
+            }
+            // do nothing if they are equal
         }
 
-        _update(balance0, balance1, _reserve0, _reserve1);
-        emit Swap(msg.sender, amount0In, amount1In, amount0Out, amount1Out, to);
+        require(_totalToken0() * _totalToken1() >= uint256(reserve0) * reserve1, "CP: K");
+        _safeTransfer(tokenOut, to, amountOut);
+        _update(_totalToken0(), _totalToken1(), _reserve0, _reserve1);
```

Comment:
You need to send the tokens they are owed before you check K.

The flow is:

1. send token
2. callback
3. check K

Realize flashloans are not implemented yet but would've thought the flow would still be the same

Reply:
yes this is a mistake... I must have been delirious. Fixed.